### ISSUE TITLE
Easier Cross-Platform Android Build for non-cmake projects (but still using android.toolchain.cmake)

### ISSUE THREAD

**Issue by choule-spiria** [conan-community/community#182]
### Description of Problem, Request, or Question
Android Core Dev Team recommends people to use the `android.toolchain.cmake` file to configure their cross-platform build needs. 

All is fine and works beautifully for `cmake`-based projects, but for other older / beefier projects, that use other build systems, we're left out cold. Even though it still works, the `make-standalone-toolchar.py` script is deprecated, so it's going to be removed eventually. (and let's be honest, it is a pain to configure in the conan profile.)

The `android.toolchain.cmake` already contains all we need, we just need to extract it, and pass it to the underlying build system (which is not cmake)

I experimented with cmake a bit, and I found out that you can print out the list of cmake variables for a given project, with the following command:

```
cmake . \
-DANDROID_ABI=armeabi-v7a \
-DANDROID_NDK=$HOME/Library/Android/sdk/ndk-bundle \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_TOOLCHAIN_FILE=$HOME/Library/Android/sdk/ndk-bundle/build/cmake/android.toolchain.cmake \
-DANDROID_TOOLCHAIN=clang \
-DANDROID_NATIVE_API_LEVEL=21 -LA
```

Resulting outputs contains something like:

```
CMAKE_AR:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ar
CMAKE_ASM_FLAGS:STRING=
CMAKE_ASM_FLAGS_DEBUG:STRING=
CMAKE_ASM_FLAGS_RELEASE:STRING=
CMAKE_COLOR_MAKEFILE:BOOL=ON
CMAKE_CXX_COMPILER_AR:FILEPATH=CMAKE_CXX_COMPILER_AR-NOTFOUND
CMAKE_CXX_COMPILER_RANLIB:FILEPATH=CMAKE_CXX_COMPILER_RANLIB-NOTFOUND
CMAKE_CXX_FLAGS:STRING=
CMAKE_CXX_FLAGS_DEBUG:STRING=
CMAKE_CXX_FLAGS_MINSIZEREL:STRING=-Os -DNDEBUG
CMAKE_CXX_FLAGS_RELEASE:STRING=
CMAKE_CXX_FLAGS_RELWITHDEBINFO:STRING=-O2 -g -DNDEBUG
CMAKE_CXX_STANDARD_LIBRARIES:STRING=-latomic -lm
CMAKE_C_COMPILER_AR:FILEPATH=CMAKE_C_COMPILER_AR-NOTFOUND
CMAKE_C_COMPILER_RANLIB:FILEPATH=CMAKE_C_COMPILER_RANLIB-NOTFOUND
CMAKE_C_FLAGS:STRING=
CMAKE_C_FLAGS_DEBUG:STRING=
CMAKE_C_FLAGS_MINSIZEREL:STRING=-Os -DNDEBUG
CMAKE_C_FLAGS_RELEASE:STRING=
CMAKE_C_FLAGS_RELWITHDEBINFO:STRING=-O2 -g -DNDEBUG
CMAKE_C_STANDARD_LIBRARIES:STRING=-latomic -lm
CMAKE_EXE_LINKER_FLAGS:STRING=
CMAKE_EXE_LINKER_FLAGS_DEBUG:STRING=
CMAKE_EXE_LINKER_FLAGS_MINSIZEREL:STRING=
CMAKE_EXE_LINKER_FLAGS_RELEASE:STRING=
CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_EXPORT_COMPILE_COMMANDS:BOOL=OFF
CMAKE_INSTALL_PREFIX:PATH=/usr/local
CMAKE_LINKER:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ld
CMAKE_MAKE_PROGRAM:FILEPATH=/usr/bin/make
CMAKE_MODULE_LINKER_FLAGS:STRING=
CMAKE_MODULE_LINKER_FLAGS_DEBUG:STRING=
CMAKE_MODULE_LINKER_FLAGS_MINSIZEREL:STRING=
CMAKE_MODULE_LINKER_FLAGS_RELEASE:STRING=
CMAKE_MODULE_LINKER_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_NM:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-nm
CMAKE_OBJCOPY:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-objcopy
CMAKE_OBJDUMP:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-objdump
CMAKE_RANLIB:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ranlib
CMAKE_SHARED_LINKER_FLAGS:STRING=
CMAKE_SHARED_LINKER_FLAGS_DEBUG:STRING=
CMAKE_SHARED_LINKER_FLAGS_MINSIZEREL:STRING=
CMAKE_SHARED_LINKER_FLAGS_RELEASE:STRING=
CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_SKIP_INSTALL_RPATH:BOOL=NO
CMAKE_SKIP_RPATH:BOOL=NO
CMAKE_STATIC_LINKER_FLAGS:STRING=
CMAKE_STATIC_LINKER_FLAGS_DEBUG:STRING=
CMAKE_STATIC_LINKER_FLAGS_MINSIZEREL:STRING=
CMAKE_STATIC_LINKER_FLAGS_RELEASE:STRING=
CMAKE_STATIC_LINKER_FLAGS_RELWITHDEBINFO:STRING=
CMAKE_STRIP:FILEPATH=$HOME/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-strip
CMAKE_VERBOSE_MAKEFILE:BOOL=FALSE
```

### Package Details (Include if Applicable)
* Package Name/Version: **boost/1.69.0**
* Operating System: **MacOS**
* Operation System Version: **10.14.4**
* Compiler+version: **clang-8**
* Conan version: **conan 1.14.1**

### Steps to reproduce (Include if Applicable)
* Checkout https://github.com/conan-community/conan-boost
* Use a simplistic Android conan profile like
```
include(default)

[settings]
os=Android
os.api_level=21
arch=armv7
compiler=clang
compiler.version=8
compiler.libcxx=libstdc++
build_type=Release

[env]
CONAN_CMAKE_TOOLCHAIN_FILE=$HOME/Library/Android/sdk/ndk-bundle/build/cmake/android.toolchain.cmake
```

* go in the conan directory and call `conan create . conan/stable --profile=/path/to/the/profile/android_armv7 --build missing`

### Build logs (Include if Available)
```
boost/1.69.0@conan/stable: Generator txt created conanbuildinfo.txt
boost/1.69.0@conan/stable: Calling build()
boost/1.69.0@conan/stable: ./bootstrap.sh
-n Bootstrapping the build engine with toolset darwin...
engine/bin.macosxx86_64/b2

Bootstrapping is done. To build and install, run:

    ./b2 install --prefix=<DIR>

boost/1.69.0@conan/stable: Cross building, detecting compiler...
boost/1.69.0@conan/stable: Cross building flags: []
boost/1.69.0@conan/stable: WARN: Patching user-config.jam
boost/1.69.0@conan/stable:
boost/1.69.0@conan/stable: ERROR: Package '353a5f8cca00431c64c6616b463bd1cdd7c4180f' build failed
boost/1.69.0@conan/stable: WARN: Build folder ~/.conan/data/boost/1.69.0/conan/stable/build/353a5f8cca00431c64c6616b463bd1cdd7c4180f
ERROR: boost/1.69.0@conan/stable: Error in build() method, line 398
	self.create_user_config_jam(self._boost_build_dir)
while calling 'create_user_config_jam', line 681
	contents += ' "%s"' % exe.replace("\\", "/")
	AttributeError: 'NoneType' object has no attribute 'replace'
```

**Comment by SSE4** [comment]
hi @choule-spiria 
two things
1. there was a bug in recipe, when `exe` was set to None, I think I have fixed it now: https://github.com/conan-community/conan-boost/commit/a4248984caa05d519911f2e153e1806a009fb861
2. for using everything from the `android.toolchain.cmake`, I think it's possible to create a small script which will either generate conan profile from it, or just set required environment variables like CFLAGS, ect. WDYT? I'd like to avoid putting this logic into recipe, as a) it's needed not for just a single recipe, so it doesn't scale well putting this code into recipes b) recipe is already pretty complex

**Comment by limbolily** [comment]
How about enhance the conan.CMake class with the ability to automatically recognize the profile with Android settings and then generate appropriate cmake definitions.

**Comment by SSE4** [comment]
@limbolily yeah, it might be good. have you seen this one? https://github.com/conan-io/conan/pull/4538

**Comment by limbolily** [comment]
@SSE4 That's cool,look forward to use it in next stable version.

**Comment by SSE4** [comment]
@limbolily should be available in conan 1.16 release.

**Comment by SSE4** [comment]
original problem was solved, I am closing an issue for now. please open new issue(s) for another problems.
